hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java

```diff
@@ -0,0 +1,269 @@
+
+/**
+ * Created by jyuan on 2/12/18.
+ */
+public class SparkTableChecker implements TableChecker {
+
+    protected static final Logger LOG=Logger.getLogger(SparkTableChecker.class);
+
+    private long baseTableCount = 0;
+    private long indexCount = 0;
+    private long invalidIndexCount = 0;
+    private long missingIndexCount = 0;
+    private String schemaName;
+    private String tableName;
+    private String indexName;
+    private DataSet<ExecRow> baseTable;
+    private PairDataSet<String, byte[]> filteredTable;
+    private KeyHashDecoder tableKeyDecoder;
+    private ExecRow tableKeyTemplate;
+    private KeyHashDecoder indexKeyDecoder;
+    private ExecRow indexKeyTemplate;
+    private int  maxCheckTableError;
+    private LeadingIndexColumnInfo leadingIndexColumnInfo;
+    private long filteredTableCount;
+
+    public SparkTableChecker(String schemaName,
+                             String tableName,
+                             DataSet table,
+                             KeyHashDecoder tableKeyDecoder,
+                             ExecRow tableKey) {
+        this.schemaName = schemaName;
+        this.tableName = tableName;
+        this.baseTable = table;
+        this.tableKeyDecoder = tableKeyDecoder;
+        this.tableKeyTemplate = tableKey;
+        maxCheckTableError = SIDriver.driver().getConfiguration().getMaxCheckTableErrors();
+    }
+
+    @Override
+    public List<String> checkIndex(PairDataSet index,
+                                   String indexName,
+                                   KeyHashDecoder indexKeyDecoder,
+                                   ExecRow indexKey,
+                                   LeadingIndexColumnInfo leadingIndexColumnInfo) throws Exception {
+
+        this.indexName = indexName;
+        this.indexKeyDecoder = indexKeyDecoder;
+        this.indexKeyTemplate = indexKey;
+        this.leadingIndexColumnInfo = leadingIndexColumnInfo;
+
+        List<String> messages = Lists.newLinkedList();
+
+        // Count number of rows in base table and reuse it if the index does not exclude default or null keys
+        JavaFutureAction<Long> tableCountFuture = null;
+        filteredTable = baseTable.filter(new IndexFilter<>(leadingIndexColumnInfo)).index(new KeyByRowIdFunction<>());
+        if (baseTableCount == 0 || leadingIndexColumnInfo != null) {
+            SpliceSpark.pushScope(String.format("Count table %s.%s", schemaName, tableName));
+            tableCountFuture = ((SparkPairDataSet) filteredTable).rdd.countAsync();
+            SpliceSpark.popScope();
+        }
+        // count number of rows in the index
+        SpliceSpark.pushScope(String.format("Count index %s.%s", schemaName, indexName));
+        JavaFutureAction<Long> indexCountFuture = ((SparkPairDataSet)index).rdd.countAsync();
+        SpliceSpark.popScope();
+
+        messages.addAll(checkMissingIndexes(filteredTable, index));
+        messages.addAll(checkInvalidIndexes(filteredTable, index));
```

Comment:
I think the call of checkInvalidIndexes() can be moved after getting the rowcount from table and index. As if rowcounts of table and index are the same and checkMissingIndexes() does not return any error, we don't need to call checkInvalidIndexes().

Reply:
checkInvalidIndexes() always needs to run even if the counts matches and no missing or duplicate index.